sources/api/migration/migrator/src/main.rs

```diff
@@ -580,60 +783,300 @@ where
-        let m04_2 = Path::new("migrate_v0.0.4_002");
-        let all_migrations = vec![&m00_1, &m01_1, &m01_2, &m02_1, &m03_1, &m04_1, &m04_2];
-
-        // Versions for use in testing
-        let v00 = Version::new(0, 0, 0);
-        let v01 = Version::new(0, 0, 1);
-        let v02 = Version::new(0, 0, 2);
-        let v03 = Version::new(0, 0, 3);
-        let v04 = Version::new(0, 0, 4);
-        let v05 = Version::new(0, 0, 5);
-
-        // Test going forward one minor version
-        assert_eq!(
-            select_migrations(&v01, &v02, &all_migrations).unwrap(),
-            vec![m02_1]
-        );
+    /// Returns the filepath to a `root.json` file stored in tree for testing. This file declares
+    /// an expiration date of `1970-01-01` to ensure success with an expired TUF repository.
+    fn root() -> PathBuf {
+        test_data()
+            .join("expired-root.json")
+            .canonicalize()
+            .unwrap()
+    }
 
-        // Test going backward one minor version
-        assert_eq!(
-            select_migrations(&v02, &v01, &all_migrations).unwrap(),
-            vec![m02_1]
-        );
+    /// Returns the filepath to a private key, stored in tree and used only for testing.
+    fn pem() -> PathBuf {
+        test_data().join("snakeoil.pem").canonicalize().unwrap()
+    }
 
-        // Test going forward a few minor versions
-        assert_eq!(
-            select_migrations(&v01, &v04, &all_migrations).unwrap(),
-            vec![m02_1, m03_1, m04_1, m04_2]
-        );
+    /// Returns the filepath to a Rust file that defines a small test migration program.
+    fn migration_rs() -> PathBuf {
+        test_data().join("migration.rs").canonicalize().unwrap()
+    }
 
-        // Test going backward a few minor versions
-        assert_eq!(
-            select_migrations(&v04, &v01, &all_migrations).unwrap(),
-            vec![m04_2, m04_1, m03_1, m02_1]
-        );
+    /// The name of a test migration. The prefix `b-` ensures we are not alphabetically sorting.
+    const FIRST_MIGRATION: &str = "b-first-migration";
+
+    /// The name of a test migration. The prefix `a-` ensures we are not alphabetically sorting.
+    const SECOND_MIGRATION: &str = "a-second-migration";
 
-        // Test no matching migrations
-        assert!(select_migrations(&v04, &v05, &all_migrations)
+    /// Creates a migration executable binary by compiling a small Rust main program. Returns the
+    /// binary as bytes.
+    fn create_test_migration<S: AsRef<str>>(migration_name: S) -> Vec<u8> {
+        let sourcecode = std::fs::read_to_string(migration_rs())
             .unwrap()
-            .is_empty());
+            .replace("migration-name-replaceme", migration_name.as_ref());
+        let tempdir = TempDir::new().unwrap();
+        let source_file = tempdir.path().join("migration.rs");
+        std::fs::write(&source_file, sourcecode.as_bytes()).unwrap();
+        let compile_command_result = Command::new("rustc")
```

Comment:
I have strong reservations about calling rustc in a test; it doesn't seem expected, reliable, performance, or safe in general.  My preferences would be (1) shell script with pentacle, which we have ideas on making work; (2) check in a binary (made properly small); (3) no test; (4) rustc.

Reply:
> webern force-pushed the webern:pr/signed-migrations branch from 1f69203 to afc4b5d 13 seconds ago

I removed the large end-to-end migrator 'unit' test. I will try to introduce it in a future PR where we can spend more time refining it.

Reply:
We have the ability to run shell scripts with pentacle 0.2, which we can use to add the test back.  I think you wanted to do that in a separate PR; resolving.